File: frontend/src/features/devops/git/types/pipelines.ts
// Git Pipeline Types

import type { PaginationInfo } from './repositories';

export interface GitPipeline {
  id: string;
  external_id: string;
  name: string;
  status: 'pending' | 'running' | 'completed' | 'cancelled';
  conclusion?: 'success' | 'failure' | 'cancelled' | 'skipped' | 'timed_out';
  trigger_event?: string;
  ref?: string;
  branch_name?: string;
  sha?: string;
  short_sha?: string;
  actor_username?: string;
  web_url?: string;
  run_number?: number;
  run_attempt?: number;
  total_jobs: number;
  completed_jobs: number;
  failed_jobs: number;
  progress_percentage: number;
  duration_seconds?: number;
  duration_formatted?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
}

export interface GitPipelineJob {
  id: string;
  external_id: string;
  name: string;
  status: 'pending' | 'running' | 'completed' | 'cancelled';
  conclusion?: string;
  step_number?: number;
  runner_name?: string;
  runner_os?: string;
  duration_seconds?: number;
  duration_formatted?: string;
  logs_available: boolean;
  completed_steps?: number;
  total_steps?: number;
  started_at?: string;
  completed_at?: string;
  created_at: string;
}

export interface PipelineStep {
  name: string;
  status: string;
  conclusion?: string;
  number: number;
}

export interface GitPipelineDetail extends GitPipeline {
  jobs: GitPipelineJob[];
  workflow_config?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface GitPipelineJobDetail extends GitPipelineJob {
  logs_url?: string;
  steps?: PipelineStep[];
  outputs?: Record<string, unknown>;
  runner_info?: Record<string, unknown>;
  act_runner?: boolean;
}

export interface PipelineStats {
  total_runs: number;
  success_count: number;
  failed_count: number;
  cancelled_count: number;
  success_rate: number;
  avg_duration_seconds: number;
  runs_today: number;
  runs_this_week: number;
  active_runs: number;
}

export interface GitPipelinesResponse {
  pipelines: GitPipeline[];
  pagination: PaginationInfo;
  stats: PipelineStats;
}

File: server/spec/factories/git/pipelines.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :git_pipeline, class: 'Devops::GitPipeline' do
    association :repository, factory: :git_repository
    association :account

    sequence(:external_id) { |n| "pipeline#{n}" }
    sequence(:name) { |n| "CI Pipeline #{n}" }
    status { 'pending' }
    conclusion { nil }
    trigger_event { 'push' }
    ref { 'refs/heads/main' }
    sha { SecureRandom.hex(20) }
    actor_username { 'testuser' }
    web_url { nil }
    logs_url { nil }
    sequence(:run_number) { |n| n }
    run_attempt { 1 }
    total_jobs { 3 }
    completed_jobs { 0 }
    failed_jobs { 0 }
    duration_seconds { nil }
    workflow_config { {} }
    started_at { nil }
    completed_at { nil }

    trait :pending do
      status { 'pending' }
      conclusion { nil }
    end

    trait :running do
      status { 'in_progress' }
      conclusion { nil }
      started_at { 5.minutes.ago }
      completed_jobs { 1 }
    end

    trait :completed do
      status { 'completed' }
      conclusion { 'success' }
      started_at { 10.minutes.ago }
      completed_at { 5.minutes.ago }
      completed_jobs { 3 }
      duration_seconds { 300 }
    end

    trait :success do
      status { 'completed' }
      conclusion { 'success' }
      started_at { 10.minutes.ago }
      completed_at { 5.minutes.ago }
      completed_jobs { 3 }
      failed_jobs { 0 }
      duration_seconds { 300 }
    end

    trait :failure do
      status { 'completed' }
      conclusion { 'failure' }
      started_at { 10.minutes.ago }
      completed_at { 5.minutes.ago }
      completed_jobs { 2 }
      failed_jobs { 1 }
      duration_seconds { 180 }
    end

    trait :cancelled do
      status { 'completed' }
      conclusion { 'cancelled' }
      started_at { 10.minutes.ago }
      completed_at { 3.minutes.ago }
      completed_jobs { 1 }
      duration_seconds { 420 }
    end

    trait :with_workflow_config do
      workflow_config do
        {
          file: '.github/workflows/ci.yml',
          name: 'CI',
          on: { push: { branches: [ 'main' ] } }
        }
      end
    end
  end
end
